### PR TITLE
sql/pgwire: send placeholder BackendKeyData

### LINKED ISSUE #13191: pgwire: python3 py-postgresql driver error
Tried to connect using py-postgresql driver
```
postgresql.exceptions.ClientCannotConnectError: could not establish connection to server
  CODE: 08001
  LOCATION: CLIENT
CONNECTION: [failed]
  failures[0]:
    NOSSL socket('192.168.8.170', 26257)
    postgresql.exceptions.ProtocolError: expected message of types b'K', but received b'Z' instead
      CODE: 08P01
      LOCATION: CLIENT
CONNECTOR: [IP4] pq://root@192.168.8.170:26257
  category: None
DRIVER: postgresql.driver.pq3.Driver
```

### PR DESCRIPTION
**Pull request by rafiss**
fixes #13191 

Some tools expect this message to be returned at connection time and
will not connect without it. CockroachDB does not support pgwire
cancellation, but we can still send a placeholder value here, and
continue ignoring cancellation requests like we already do.

Added a small test to make sure nothing broke.

Release note (sql change): When a connection is established, CockroachDB
will now return a placeholder BackendKeyData message in the response.
This is for compatibility with some tools, but using the BackendKeyData
to cancel a query will still have no effect, just the same as before.

### CODE REVIEW

**Comment by rafiss** [comment]
@asubiotto @jordanlewis i've requested your review since you may have more context from #13009

**Review by jordanlewis — COMMENTED** [review]
Can you add some pgwire data driven tests for this too? I'm actually surprised we didn't see any failures in those, since we're changing what we're sending - what explains that?

What ORM/driver does this change target, again? Do we have tests for it?

Overall I think this looks good. Have you already gone spelunking for why we didn't merge a placeholder BackendKeyData originally? I vaguely recall @asubiotto sending a PR for this, but it obviously never got merged and I wonder why that's the case.

**Comment by rafiss** [comment]
> Can you add some pgwire data driven tests for this too? I'm actually surprised we didn't see any failures in those, since we're changing what we're sending - what explains that?

i think it's because this message is during the connection handshake, and those tests only include assertions after the connection has been established. i'll look into enhancing those tests.

> What ORM/driver does this change target, again? Do we have tests for it?

the tool mentioned in #13191 (which is the issue that prompted my attention here) is the py-postgresql driver for python3. the tool mentioned in #13009 (which was Alfonso's earlier PR) is pgpool. we don't have tests for these tools. i'm not sure if we should, since our approach has only been to test tools that we want to document support for, but i'm open to hearing an argument in favor of testing now.

> Overall I think this looks good. Have you already gone spelunking for why we didn't merge a placeholder BackendKeyData originally? I vaguely recall @asubiotto sending a PR for this, but it obviously never got merged and I wonder why that's the case.

Alfonso's comment (https://github.com/cockroachdb/cockroach/pull/13009#issuecomment-336956657) describes why. at that time, it seems that we were waiting until there was telemetry and unimplemented errors for when we received a CancelRequest message. we now do have this in place.

**Comment by jordanlewis** [comment]
Ok, sounds good. We need to have at least some test that this functionality is actually usable by clients - that they receive the key data without larger issues. I suppose all of the ORM tests will check this automatically too.

LGTM if you can check via the pgwire datadriven tests (or really any pgwire test that actually inspects the bytes) that we're sending the cancel data message - would be a shame if we had a regression here and never noticed.

**Review by asubiotto — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/pgwire/conn.go, line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdjDKBqCp8hQhRNum:-MTAdjDKBqCp8hQhRNun:bljly4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn.go#L310)):*
> ```Go
> 		// normally used by clients to send a CancelRequest message:
> 		// https://www.postgresql.org/docs/9.6/static/protocol-flow.html#AEN112861
> 		// CockroachDB currently ignores all CancelRequests.
> ```

Is this actually ignored or do we return some unimplemented error? If the former, would it make sense/be a lot of work to return an error? OK to do this work separately.
___
*[pkg/sql/pgwire/conn_test.go, line 1525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAd2PHBAqHBMj17bpV:-MTAd2PHBAqHBMj17bpW:bxrz82t) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn_test.go#L1525)):*
> ```Go
> }
> 
> // TestReadTimeoutConn asserts that a readTimeoutConn performs reads normally
> ```

nit: name
___
*[pkg/sql/pgwire/conn_test.go, line 1545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdIrd8JsP-gATzHzf:-MTAdIre2nmc9EeIGcaS:b4n5c16) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn_test.go#L1545)):*
> ```Go
> 	ctx, cancel := context.WithCancel(context.Background())
> 
> 	// Delay execution for just a bit until db.QueryContext has begun.
> ```

Hmm, this looks like it's going to be flaky. You could use one of the executor testing knob callbacks (Before{Prepare,Execute}) in `StartServer` to only cancel the context once it receives the `select pg_sleep` statement on the server side.

**Comment by asubiotto** [comment]
Thanks for this change BTW! Good to close such an early issue.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/pgwire/conn.go, line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdjDKBqCp8hQhRNum:-MTBW3UGAyFtftO1fdRS:b365mb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn.go#L310)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this actually ignored or do we return some unimplemented error? If the former, would it make sense/be a lot of work to return an error? OK to do this work separately.
</blockquote></details>

we don't return an error, since it is actually quite common for clients to call already. we do have telemetry on it:

https://github.com/cockroachdb/cockroach/blob/cd39473974bfb0f8b0a886ed8246e6b887f2b851/pkg/sql/pgwire/server.go#L672-L678

the data here shows that tools frequently invoke this, so it could potentially have far-reaching effects to change this to be an error: https://github.com/cockroachdb/cockroach/issues/41335#issuecomment-669658073
___
*[pkg/sql/pgwire/conn_test.go, line 1525 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAd2PHBAqHBMj17bpV:-MTBVYDq-HaZek4gDfpy:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn_test.go#L1525)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: name
</blockquote></details>

fixed
___
*[pkg/sql/pgwire/conn_test.go, line 1545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdIrd8JsP-gATzHzf:-MTBVh4JAMdiD-tZvAIO:b-ushszo) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn_test.go#L1545)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Hmm, this looks like it's going to be flaky. You could use one of the executor testing knob callbacks (Before{Prepare,Execute}) in `StartServer` to only cancel the context once it receives the `select pg_sleep` statement on the server side.
</blockquote></details>

nice! thanks for this pointer. for some context, i just copied this test from the lib/pq test suite: https://github.com/lib/pq/blob/4604d39ddc9f62e2ca152114c73aec99b77a3468/go18_test.go#L116

using callbacks is way better though

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/pgwire/conn.go, line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdjDKBqCp8hQhRNum:-MTBWyOh1xsp5AV2II3z:ba1wlgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn.go#L310)):*
the Postgres docs make CancelRequest seem like a "best-effort" thing so, i think it's dine to not have an error even though we don't do anything

> A CancelRequest message will be ignored unless it contains the same key data (PID and secret key) passed to the frontend during connection start-up. If the request matches the PID and secret key for a currently executing backend, the processing of the current query is aborted. (In the existing implementation, this is done by sending a special signal to the backend process that is processing the query.)
> 
> The cancellation signal might or might not have any effect — for example, if it arrives after the backend has finished processing the query, then it will have no effect. If the cancellation is effective, it results in the current command being terminated early with an error message.
___
*[pkg/sql/pgwire/conn_test.go, line 1545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTAdIrd8JsP-gATzHzf:-MTBWoDT3lqQN80Y1fry:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/71399b08460a0bb59549f7e72b2481a373b5a668/pkg/sql/pgwire/conn_test.go#L1545)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice! thanks for this pointer. for some context, i just copied this test from the lib/pq test suite: https://github.com/lib/pq/blob/4604d39ddc9f62e2ca152114c73aec99b77a3468/go18_test.go#L116

using callbacks is way better though
</blockquote></details>

Done.

**Comment by rafiss** [comment]
this is ready for another look. i've added a test to `pkg/testutils/pgtest/pgtest.go` that shows the BackendKeyData being received.

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-:-MTC-Ev51Q1-XxdJXFrq:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/sql/pgwire/conn_test.go, line 1557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTC-9in9nsbCThTo6lw:-MTC-9in9nsbCThTo6lx:b-ud8i9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/24adfee4e133383b2ee6f0ca23c503f7ef943f94/pkg/sql/pgwire/conn_test.go#L1557)):*
> ```Go
> 
> 	// Cancellation has no effect on ongoing query.
> 	if _, err := db.QueryContext(cancelCtx, "select pg_sleep(0.2)"); err != nil {
> ```

nit: ok to make this 0?

**Comment by rafiss** [comment]
it looks like i need to update sqlproxy to be aware of BackendKeyData. gonna clear this with the CC team.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/pgwire/conn_test.go, line 1557 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-MTC-9in9nsbCThTo6lw:-MTC2_PDC1odWNoU6YtX:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/24adfee4e133383b2ee6f0ca23c503f7ef943f94/pkg/sql/pgwire/conn_test.go#L1557)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: ok to make this 0?
</blockquote></details>

done

**Comment by rafiss** [comment]
@spaskob i see that this affects an area of sqlproxy that you worked on. can you please take a look at the changes? also, can you confirm that this update to sqlproxy will be able to be deployed before CockroachDB 21.1 is released?

**Comment by spaskob** [comment]
Thanks @rafiss for fixing the sqlproxy code. The change looks good to me. Do you mind running https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/sqlproxyccl/run_proxy.sh to make sure the proxy works e2e?

**Comment by rafiss** [comment]
@spaskob thanks! that test passes for me locally.

how often does sqlproxy get released? as long as it's released before CC uses CRDB 21.1 we should be good.

**Comment by rafiss** [comment]
thanks for the reviews all!

bors r=asubiotto,spaskob

**Review by spaskob — COMMENTED** [review]
There's no release schedule I will pick up this commit after you merge and create a ticket to release it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-:-MTCz1Hl2SKfTSFjHMf2:ba6f4ga)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @jordanlewis)

**Review by spaskob — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60281#-:-MTCzCfv4N4VX3tJoRZi:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @jordanlewis)

**Pull request merged**
